### ISSUE TITLE
Can not import the library

### ISSUE THREAD

**Issue by janwirth** [Borewit/music-metadata#1357]
**Bug description**
I can not import the package from node. I spent 10h.

**Expected behavior**
The package is imported and I can use the `parseFile` function.

**Audio file demonstrating the problem**
~/music-metadata   
❯ npm i music-metadata

added 17 packages, and audited 18 packages in 1s

8 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```                                              
~/music-metadata-test  
❯ node
Welcome to Node.js v18.4.0.
Type ".help" for more information.
> require('music-metadata')
Uncaught:
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/janwirth/music-metadata/node_modules/music-metadata/lib/index.js not supported.
Instead change the require of index.js in null to a dynamic import() which is available in all CommonJS modules.
    at __node_internal_captureLargerStackTrace (node:internal/errors:477:5)
    at new NodeError (node:internal/errors:388:5)
    at Module._extensions..js (node:internal/modules/cjs/loader:1139:19)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Module._load (node:internal/modules/cjs/loader:834:12)
    at Module.require (node:internal/modules/cjs/loader:1012:19)
    at require (node:internal/modules/cjs/helpers:102:18) {
  code: 'ERR_REQUIRE_ESM'
}
```

**Comment by janwirth** [comment]
Downgrading to @7 solved it for me.

**Comment by monstermichl** [comment]
The problem seems to be that the module was changed to a pure ECMAScript Module (ESM) as you can read here: https://www.npmjs.com/package/music-metadata#user-content-compatibility

**Comment by Borewit** [comment]
Indeed, version 8 switched from CJS to ESM as [mentioned in the v8 change log](https://github.com/Borewit/music-metadata/releases/tag/v8.0.0).
Duplicate of #1218.

**Comment by janwirth** [comment]
Yes, I was really struggling to get it to work with various combinations of transpilation, typescript configurations etc. It would be useful to have a guide on how to set it up. I am running Node 16.15.0 on ts-node.

**Comment by monstermichl** [comment]
Here's a nice guide: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

**Comment by janwirth** [comment]
I was not using `require` before, but import. I am using this template here: https://github.com/electron-react-boilerplate/electron-react-boilerplate

**Comment by janwirth** [comment]
I don't usually get stuck on stuff like this - I have several years of experience with JS and I'm familiar with all the .mjs, .cjs, type: module, --experimental-modules, magic webpack comments etc. you can do but here I was really out of my luck.

EDIT: I will revisit this with the ts-node guide that I saw in the link to the guide you sent. Thank you.

**Comment by Borewit** [comment]
The lack of backward compatibility in ESM modules is a true nightmare @janwirth.

**Comment by martpie** [comment]
@Borewit Just faced the same issue, are you sure the ESM exports are matching the CJS exports from v7? It seems like `parseFile` is not exposed.

```
"parseFile" is not exported by "node_modules/music-metadata/lib/core.js", imported by "src/main/modules/ipc-library.ts".
```

content of `import * as mmd`:
```
[Object: null prototype] [Module] {
  parseStream: [Function: parseStream],
  parseBuffer: [AsyncFunction: parseBuffer],
  parseFromTokenizer: [Function: parseFromTokenizer],
  orderTags: [Function: orderTags],
  ratingToStars: [Function: ratingToStars],
  selectCover: [Function: selectCover],
  scanAppendingHeaders: [AsyncFunction: scanAppendingHeaders]
}
```

I am using Vite, not sure if it is related.

**Comment by stuartambient** [comment]
I'm using Vite with Electron.  
```
Instead change the require of C:\Users\user\Documents\nodeProjs\musicplayer-electron\node_modules\music-metadata\lib\index.js in C:\Users\sambi\Documents\nodeProjs\musicplayer-electron\out\main\index.js to a dynamic import() which is available in all CommonJS modules.
    at c._load (node:electron/js2c/asar_bundle:5:13343)
```

I believe that while Electron doesn't support ESM ,  the Vite - Electron set up I ran does [@create-electron](https://github.com/alex8088/quick-start/tree/master/packages/create-electron)

Not sure what my next step is here,

**Comment by Borewit** [comment]
> For now I'll downgrade to 7.13 , the Vite CJS -> ESM seems to be working for that version.

[music-metadata 7.13](https://github.com/Borewit/music-metadata/releases/tag/v7.13.0) is CommonJS, not ESM as [already discussed](https://github.com/Borewit/music-metadata/issues/1357#issuecomment-1288070303)

Electron ES Module support is still an [open issue](https://github.com/electron/electron/issues/21457) at this point in time.

**Comment by stuartambient** [comment]
> > For now I'll downgrade to 7.13 , the Vite CJS -> ESM seems to be working for that version.
> 
> [music-metadata 7.13](https://github.com/Borewit/music-metadata/releases/tag/v7.13.0) is CommonJS, not ESM as [already discussed](https://github.com/Borewit/music-metadata/issues/1357#issuecomment-1288070303)
> 
> Electron ES Module support is still an [open issue](https://github.com/electron/electron/issues/21457) at this point in time.

I know and using the Electron-Vite template converts CJS to ESM.  So I can use ```import {parseFile} from 'music-metadata' ``` event though it is v7.13

**Comment by StarrKiss** [comment]
I am having this same issue but with import - this is running through electron. I'm importing it like so:
`import { parseFile } from 'music-metadata';`

and yet I get this error:
`Uncaught TypeError: Failed to resolve module specifier "music-metadata". Relative references must start with either "/", "./", or "../".`
I have absolutely no clue what the problem is.

**Comment by martpie** [comment]
yeah, `parseFile` is not available from the default export in v8, weirdly

**Comment by Borewit** [comment]
Gents, unfortunately I do not have the time to try and test every framework. 

I do not mind you use issue here to solve integration challenges, and share recommendations, but in my opinion the issues you describe are not caused by [music-metadata](https://github.com/Borewit/music-metadata).

I offer both CJS (in version 7) and ESM support (in version 8) module at the moment, a time consuming compensation from my end for the poor backward compatibility and lack of ESM clearly breaking taking over. Pick the one you need.

In the ESM module, `parseFile()` is only exported to Node, as it depends on Node specific API's. Client (browser) side typically follows the non-node path:

That distinction is made here, and is my opinion perfectly valid: 
https://github.com/Borewit/music-metadata/blob/ab248fb040cff30401c12767b3819c17b71e6115/package.json#L14-L19

Regarding using music-metadata in Electron, after having almost the same discussion, see [this post](https://github.com/Borewit/music-metadata/issues/1425#issuecomment-1356877483)

**Comment by martpie** [comment]
Thanks @Borewit for the details! I'll look deeper into that see what is going wrong exactly for Node.js, thanks for the code pointers!

I'll open a PR if needed :)

**Comment by csimpi** [comment]
@Borewit 
I use it from a very basic typescript project and the same is happening, I cannot use it and can't figure out what's going on.

Basic typescript project, running from the command line.

```

[INFO] 00:41:07 ts-node-dev ver. 2.0.0 (using ts-node ver. 10.9.1, typescript ver. 4.9.4)
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /music-match/node_modules/music-metadata/lib/index.js
require() of ES modules is not supported.
require() of /music-match/node_modules/music-metadata/lib/index.js from /music-match/src/music-match.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /music-match/node_modules/music-metadata/package.json.

    at new NodeError (internal/errors.js:322:7)
    at Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
    at require.extensions..jsx.require.extensions..js (/tmp/ts-node-dev-hook-9363314290809173.js:114:20)
    at Object.nodeDevHook [as .js] (/music-match/node_modules/ts-node-dev/lib/hook.js:63:13)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:101:18)
    at Object.<anonymous> (/music-match/src/music-match.ts:8:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
[ERROR] 00:41:09 Error: Must use import to load ES Module: /music-match/node_modules/music-metadata/lib/index.js
require() of ES modules is not supported.
require() of /music-match/node_modules/music-metadata/lib/index.js from /music-match/src/music-match.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.

```

**Comment by Borewit** [comment]
@csimpi please have a look what ES Modules mean:
-  [ES modules: A cartoon deep-dive](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/)
- [JavaScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules)
- [Hello, Modules!](https://medium.com/sindre-sorhus/hello-modules-d1010b4e777b)

Looking to the error message, you are using CJS modules in your project. You have to options:
1. Easy: Use music-metadata version **7**, this is a CJS module
2. Harder: [Migrate your project to ESM](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm)

**Comment by csimpi** [comment]
@Borewit Thank you, I am using many many JS libraries with no requirement to upgrade. Can't you offer wider compatibility? Just asking, not complaining about.

**Comment by Borewit** [comment]
I offer CJS & ESM, what other compatibility do you want?
There is no difference in functionality between 7 and 8.

**Comment by csimpi** [comment]
LOL, chill buddy.
The doc didn't mention that 7 and 8 are the same just different module systems, it the doc says starting from v8 you drop CJS compatibility. Update your docs, answer less stupid questions after :)

**Comment by Borewit** [comment]
Problem with the documentation or a reader who broadcasts first and then starts reading?

**Comment by csimpi** [comment]
@Borewit 
In the case the documentation contains the information the problem with the reader, in the case the documentation doesn't contain a necessary information or it's not clear, it's the documentation :)

**Comment by julianpoemp** [comment]
> @csimpi please have a look what ES Modules mean:
> 
>     * [ES modules: A cartoon deep-dive](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/)
> 
>     * [JavaScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules)
> 
>     * [Hello, Modules!](https://medium.com/sindre-sorhus/hello-modules-d1010b4e777b)
> 
> 
> Looking to the error message, you are using CJS modules in your project. You have 2 options:
> 
>     1. Easy: Use music-metadata version **7**, this is a CJS module
> 
>     2. Harder: [Migrate your project to ESM](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm)

I don't understand why CJS support was dropped for this package. A lot of projects need CJS support and shouldn't need to use an old version (7). There are many ways for developers to build packages that support both CJS and ESM without additional overhead. For example I created many libraries using Nx with ESM and CJS support without problems. Enforcing ESM on projects is developer unfriendly.

**Comment by monstermichl** [comment]
I think the thing is that we should have moved to ESM in general quite a while ago, but it didn't really happen yet and if we keep continuing to support old stuff, people won't move to the new stuff. Human lazyness is the default behaviour :) Of course I cannot speak for the developer, maybe he had something else in mind.

**Comment by Borewit** [comment]
> I don't understand why CJS support was dropped for this package. A lot of projects need CJS support and shouldn't need to use an old version (7). There are many ways for developers to build packages that support both CJS and ESM without additional overhead.

As far I know, hybrid support comes [with a lot of downsides](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c?permalink_comment_id=3850849#gistcomment-3850849). If you believe it simply solvable, I am open to suggestions.

> As far as I can see v7 is updated in parallel to the latest version. Compiling the package to CJS and ESM would make it even easier for the developer to maintain this project. 

I did maintain both a CJS and ESM version for a while, to give time to developers to switch. I stopped doing that since version 9. 

> I don't see any disadvantages supporting both.

The overhead of maintaining 2 different branches became a significant, with an increasing complexity, as dependencies start becoming different, on this module and all the other modules I maintain. 
Most of the time was sucked into issues raised about poly-filling and explaining difference between ESM and CJS. 

I decide, like most packages I am familiar with, to make the call and switch the only official module standard: ESM.
That also allowed me, to share the modules originally targeting to Node.js, to be used in browsers without giving up the Node.js specific functions. That allowed me archive [music-metadata-browser](https://github.com/Borewit/music-metadata-browser), which only roles was, to act as an adapter for [music-metadata](https://github.com/Borewit/music-metadata) for browser environment. Going from 3 branched to maintain, back to 1.

**Comment by julianpoemp** [comment]
@Borewit 
I read the [comment](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c?permalink_comment_id=3850849#gistcomment-3850849) and I disagree with the sentence

"If part of your codebase requires from them, and another imports the same thing, duplicated code will run/bundle." 

If a part of your codebase uses "import" and the other uses "require" statements that means that one part of your codebase still uses CJS, what is against your goal using ESM only. To avoid this problem you have to make sure that your whole codebase (incl. dependencies) imports/requires the library the same way. In your case the whole codebase may only use ESM to avoid this problem. That means all your dependencies have to support ESM, what is still not ensured in those days.

Yes, if you create a package that supports both, ESM and CJS, you have to bundle both in one package. But thanks to Ts and bundlers there is no big overhead. You have one Ts codebase and you transpile it to ESM, CJS and typings. 

> I did maintain both a CJS and ESM version for a while, to give time to developers to switch. I stopped doing that since version 9. The overhead of maintaining 2 different branches became a significant [...]

That's why my suggestion is to use a bundler that creates an hyprid package supporting CJS, ESM and typings. That way you only have to maintain only one codebase. I recommend using Nx and the [js-library](https://nx.dev/nx-api/js/generators/library) generator using vite.

> Most of the time was sucked into issues raised about poly-filling and explaining difference between ESM and CJS.

Interesting, I never had those issues with hybrid packages created by bundlers like vite.

Yes, I think ESM is much better than CJS (e.g. to reduce optimization bailouts), but the compatibility with (older) frameworks and dependencies is still a problem. In cases where a codebase can't be updated to ESM users can't use your library. I think the best tradeoff is to support both CJS and ESM as long as there is a need for it. That's why modern libraries support both. Old libraries will die the long run and the modern ones can remove CJS support as soon as CJS is wiped off completely.

All in all I respect your decision, even if I have another opinion. As a developer of a library I would want to reach as many developers as possible. Supporting both CJS and ESM doesn't mean to keep CJS alive. It means to give dependencies time to switch to ESM.

Thank you for your answer.

**Comment by Borewit** [comment]
>  If a part of your codebase uses "import" and the other uses "require" statements that means that one part of your codebase still uses CJS, what is against your goal using ESM only. To avoid this problem you have to make sure that your whole codebase (incl. dependencies) imports/requires the library the same way. In your case the whole codebase may only use ESM to avoid this problem. That means all your dependencies have to support ESM, what is still not ensured in those days.

Migrating from CommonJS is a painful transition. The least painful compatibility which is available in this entire story, is that you can use CJS dependencies in an ESM project. Ideally they are all ESM, unavoidable and not the end of the world if that is not the case.

I also read this article https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html, concluding a similar solution where a CJS module is transpiled from a single code base. 

You cannot expect that millions packages compensate in an overly complex solution to fix the unavoidable and painful transition introduced by moving towards ESM.

I am really sorry users, if it would be a reasonable thing to provide backwards compatibility for CJS, I would have love to provide it, but this is, despite some may not agree, I my opinion not the case. The only advice I can give you, migrate you project ESM, it unavoidable on the long run. Some guidance how to do that can be found [here](https://github.com/sindresorhus/meta/discussions/15).

**Comment by Borewit** [comment]
@julianpoemp, for CJS compatibility, would this proposed solution work for you https://github.com/sindresorhus/file-type/pull/665 (different repo, but if I would apply that solution to music-metadata?)

Where essentially an asynchronous loader function is added for CJS backwards compatiblity?

**Comment by julianpoemp** [comment]
@Borewit I just tested music-metadata v10.3.0 with CJS Backward Compatibility. Unfortunately it's not working. I created a minimal non-working-example: https://github.com/julianpoemp/music-metadata-ts-cjs-test/blob/main/src/main.ts

You can clone that repository, run `npm install`and `npm start`. Then you get the following error message:

````
Error parsing metadata: (0 , import_music_metadata.parseStream) is not a function
```` 

Btw, one use case of the demand of making a project with CJS is packaging a NodeJS application with [pkg](https://github.com/yao-pkg/pkg). For the last years there is no ESM support. If projects have to use pkg to create binaries of their NodeJS projects, they have to use CJS.

**Comment by Borewit** [comment]
> @Borewit I just tested music-metadata v10.3.0 with CJS Backward Compatibility. Unfortunately it's not working. I created a minimal non-working-example: https://github.com/julianpoemp/music-metadata-ts-cjs-test/blob/main/src/main.ts

To utilize the CJS entry point, which assist you loading the ESM module in a CommonJS project, please try the instruction as provided in the [README](https://github.com/Borewit/music-metadata?tab=readme-ov-file#commonjs-backward-compatibility).

> Btw, one use case of the demand of making a project with CJS is packaging a NodeJS application with [pkg](https://github.com/yao-pkg/pkg). For the last years there is no ESM support.

Looks like `pkg` is almost a decade behind.

> Another example of a framework that does not support ESM only packages: https://github.com/nestjs/nest/issues/13319. If you use this framework you'll not going to switch to another framework. You won't be using ESM only packages.

Quote from the discussion:  
> you can use ESM-only packages in CommonJS projects just fine.  

Ref: https://github.com/nestjs/nest/issues/13319#issuecomment-1995995486:

**Comment by julianpoemp** [comment]
@Borewit 

> To utilize the CJS entry point, which assist you loading the ESM module in a CommonJS project, please try the instruction as provided in the [README](https://github.com/Borewit/music-metadata?tab=readme-ov-file#commonjs-backward-compatibility).

Yes. But as dev you can't do anything about it if there is no 1:1 alternative. Btw. [65% of the projects still use CJS](https://github.com/nestjs/nest/issues/13319#issuecomment-2004202362) That's why modern libraries support CJS and ESM as I described before.

> Quote from the discussion:
>    you can use ESM-only packages in CommonJS projects just fine.
> Ref: https://github.com/nestjs/nest/issues/13319#issuecomment-1995995486:

Yes, but if you follow the referenced issue you'll see that it has it's [limitations](https://github.com/nestjs/nest/pull/8736#issuecomment-1473742551).

Can you please clone the test-repo I created earlier and show me, how to get 10.3 working? Neither the standard import nor your "CJS Backward compatibility" code works. With this example project you can reproduce the problem.

**Comment by Borewit** [comment]
> Can you please clone the test-repo I created earlier and show me, how to get 10.3 working? Neither the standard import nor your "CJS Backward compatibility" code works. With this example project you can reproduce the problem.

I think the problem you run into, is that there are no Typings defined (as mention on the [README](https://github.com/Borewit/music-metadata?tab=readme-ov-file#commonjs-backward-compatibility)) for the `loadMusicMetadata` function.

To workaround the TypeScript type check, try:

```ts
import mmCjs from 'music-metadata';
import { createReadStream } from 'node:fs';

type MusicMetadataCjs = {
  loadMusicMetadata: () => Promise<typeof import('music-metadata')>;
};

async function main() {
  try {
    // Create a readable stream from a file
    const audioStream = createReadStream('src/assets/EinsZweiDrei.wav');

    // Load the music-metadata ESM module
    const mmEsm = await (mmCjs as any as MusicMetadataCjs).loadMusicMetadata();

    // Parse the metadata from the stream
    const metadata = await mmEsm.parseStream(audioStream, { mimeType: 'audio/wav' });

    // Log the parsed metadata
    console.log(metadata);
  } catch (error) {
    console.error('Error parsing metadata:', error.message);
  }
}

main();
```

Please let me know the result.

**Comment by julianpoemp** [comment]
@Borewit it work's, thank you! It's not the most beautiful solution, but it works.

**Comment by Borewit** [comment]
> @Borewit it work's, thank you! It's not the most beautiful solution, but it works.

I had no luck including Typings for `loadMusicMetadata`. Fundamental problem is that TypeScript loads the ESM typings for CJS entry point.

Neither this hack would be needed, if TypeScript would allow you to do a dynamic import in CommonJS environment.

Ref: https://github.com/TypeStrong/ts-node/discussions/1290

**Comment by Borewit** [comment]
@julianpoemp With version [v10.3.1](https://github.com/Borewit/music-metadata/releases/tag/v10.3.1) you can update your code to:

```ts
import { loadMusicMetadata } from 'music-metadata';
import { createReadStream } from 'node:fs';

async function main() {
  try {
    // Create a readable stream from a file
    const audioStream = createReadStream('src/assets/EinsZweiDrei.wav');

    const { parseStream } = await loadMusicMetadata();

    // Parse the metadata from the stream
    const metadata = await parseStream(audioStream, { mimeType: 'audio/wav' });

    // Log the parsed metadata
    console.log(metadata);
  } catch (error) {
    console.error('Error parsing metadata:', error.message);
  }
}

main();
```

Let's see how generous the CommonJS community is: https://github.com/sponsors/Borewit
